libs/s25main/world/GameWorldView.cpp

```diff
@@ -347,12 +347,17 @@ void GameWorldView::DrawNameProductivityOverlay(const TerrainRenderer& terrainRe
             if(!no)
                 continue;
 
+            Position curPos = GetWorld().GetNodePos(pt) - offset + curOffset;
+            curPos.y -= 22;
+
             // Is object not belonging to local player?
             if(no->GetPlayer() != gwv.GetPlayerId())
+            {
+                auto* attackAidImage = LOADER.GetImageN("map_new", 20000);
```

Comment:
Try to avoid doing needless work. E.g. getting the image when you rarely need it (the condition below is mostly false)
So put this into the block guarded by the `if`

Just for context what happens in `LOADER.GetImageN("map_new", 20000);`

- Look up the Singleton object
- Construct a string-like
- Search a map
- Access a large vector (cache miss likely)
- Cast the result

Reply:
Yeah, I thought about it, but then I saw that current code for military aid for construction is doing basically the same:

```cpp
void GameWorldView::DrawConstructionAid(const MapPoint& pt, const DrawPoint& curPos)
{
    BuildingQuality bq = gwv.GetBQ(pt);
    if(bq != BuildingQuality::Nothing)
    {
        constexpr auto bqImgs = getBqImgs();
        auto* bm = LOADER.GetMapTexture(bqImgs[bq]);
        // Draw building quality icon
        bm->DrawFull(curPos);
        // Show ability to construct military buildings
        if(GetWorld().GetGGS().isEnabled(AddonId::MILITARY_AID))
        {
            if(!GetWorld().IsMilitaryBuildingNearNode(pt, gwv.GetPlayerId())
               && (bq == BuildingQuality::Hut || bq == BuildingQuality::House || bq == BuildingQuality::Castle
                   || bq == BuildingQuality::Harbor))
                LOADER.GetImageN("map_new", 20000)->DrawFull(curPos - DrawPoint(-1, bm->GetSize().y + 5));
        }
    }
}
```

so I thought that we're good here and by the power of constexpr, compiler optimization and other black magic we're good with this approach. Ideally, this icon should be loaded once for whole GameWorldView, provided that plugin is enabled, and be used straight from dedicated object field.

Reply:
Sorry? In that code every image that is loaded is also used.

Reply:
yeah but it is loaded very often nonetheless - for each construction icon which qualify as spot for military building as I understand it.